dnstap/mirroroutput.go

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2019 by Farsight Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+
+import (
+	dnstap "github.com/dnstap/golang-dnstap"
+)
+
+type mirrorOutput struct {
+	outputs []dnstap.Output
+	data    chan []byte
+	done    chan struct{}
+}
+
+func newMirrorOutput() *mirrorOutput {
+	return &mirrorOutput{
+		data: make(chan []byte, outputChannelSize),
+		done: make(chan struct{}),
+	}
+}
+
+func (mo *mirrorOutput) Add(o dnstap.Output) {
+	mo.outputs = append(mo.outputs, o)
+}
+
+func (mo *mirrorOutput) RunOutputLoop() {
+	for b := range mo.data {
+		for _, o := range mo.outputs {
```

Comment:
a slightly more performant loop might be to iterate over indices (mo.outputs[i]) instead of values, which returns a copy (https://github.com/golang/go/wiki/Range#gotchas) - but not 100% sure

Reply:
I generally prefer ranging over data for clarity, but will keep this gotcha in mind for future code.

In this case, the outputs are interfaces and thus the copy is small and a suitable alias for the original.